src/flp/types/fixedpoint_l2.rs

```diff
@@ -0,0 +1,816 @@
+                }
+                padded_chunk[2 * i + 1] = r * constant_part_multiplier;
+                r *= joint_rand[0];
+            }
+
+            range_check += g[0].call(&padded_chunk)?;
+        }
+
+        // Compute the norm of the entries and ensure that it is the same as the
+        // submitted norm. There are exactly enough bits such that a submitted
+        // norm is always a valid norm (semantically in the range [0,1]). By
+        // comparing submitted with actual, we make sure the actual norm is
+        // valid.
+        //
+        // The function to compute here (see explanatory comment at the top) is
+        //   norm(ys) = sum_{y in ys} y^2 - (2^n)*y + 2^(2n-2)
+        //
+        // This is done by the `ParallelSum` gadget `g[1]`, which evaluates the
+        // inner polynomial on each (decoded) vector entry, and then sums the
+        // results. Note that the gadget is not called on the whole vector at
+        // once, but sequentially on chunks of size `self.gadget1_chunk_len` of
+        // it. The results of these calls are accumulated in the `outp` variable.
+        //
+        // decode the bit-encoded entries into elements in the range [0,2^n):
+        let decoded_entries: Result<Vec<_>, _> = input[0..self.entries * self.bits_per_entry]
+            .chunks(self.bits_per_entry)
+            .map(F::decode_from_bitvector_representation)
+            .collect();
+
+        // run parallel sum gadget on the decoded entries
+        let computed_norm = {
+            let mut outp = F::zero();
+
+            for chunk in decoded_entries?.chunks(self.gadget1_chunk_len) {
+                let d = chunk.len();
+
+                if d == self.gadget1_chunk_len {
+                    outp += g[1].call(chunk)?;
+                } else {
+                    // If the chunk is smaller than the chunk length, extend
+                    // chunk with zeros.
+                    let mut padded_chunk: Vec<_> = chunk.to_owned();
+                    padded_chunk.resize(self.gadget1_chunk_len, F::zero());
+                    outp += g[1].call(&padded_chunk)?;
+                }
+            }
+
+            outp
+        };
+
+        // The submitted norm is also decoded from its bit-encoding, and
+        // compared with the computed norm.
+        let submitted_norm_enc = &input[self.range_norm_begin..self.range_norm_end];
+        let submitted_norm = F::decode_from_bitvector_representation(submitted_norm_enc)?;
+
+        let norm_check = computed_norm - submitted_norm;
+
+        // Finally, we require both checks to be successful by computing a
+        // random linear combination of them.
+        let out = joint_rand[1] * range_check + (joint_rand[1] * joint_rand[1]) * norm_check;
+        Ok(out)
+    }
+
+    fn truncate(&self, input: Vec<F>) -> Result<Vec<Self::Field>, FlpError> {
+        self.truncate_call_check(&input)?;
+
+        let mut decoded_vector = vec![];
```

Comment:
You can help the compiler save mallocs by predicting the size of buffer you'll need.
```suggestion
        let mut decoded_vector = Vec::with_capacity(self.entries);
```